src/EditorFeatures/Core.Wpf/Preview/PreviewFactoryService.cs

```diff
@@ -315,46 +362,51 @@ public Task<object> CreateRemovedDocumentPreviewViewAsync(Document document, dou
         }
 
-        public Task<object> CreateRemovedAdditionalDocumentPreviewViewAsync(TextDocument document, double zoomLevel, CancellationToken cancellationToken)
+        public Task<object> CreateRemovedDocumentPreviewViewAsync(Document document, double zoomLevel, CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            // Note: We don't use the original buffer that is associated with oldDocument
-            // (and possibly open in the editor) for oldBuffer below. This is because oldBuffer
-            // will be used inside a projection buffer inside our inline diff preview below
-            // and platform's implementation currently has a bug where projection buffers
-            // are being leaked. This leak means that if we use the original buffer that is
-            // currently visible in the editor here, the projection buffer span calculation
-            // would be triggered every time user changes some code in this buffer (even though
-            // the diff view would long have been dismissed by the time user edits the code)
-            // resulting in crashes. Instead we create a new buffer from the same content.
-            // TODO: We could use ITextBufferCloneService instead here to clone the original buffer.
-            var oldBuffer = CreateNewPlainTextBuffer(document, cancellationToken);
+            return CreateRemovedTextDocumentPreviewViewAsync(
+                document, zoomLevel,
+                createBuffer: (textDocument, cancellationToken) => CreateNewBuffer((Document)textDocument, cancellationToken),
+                removeTextDocument: (solution, documentId) => solution.RemoveDocument(documentId),
+                addTextDocument: (solution, documentId, name, text) => solution.AddDocument(documentId, name, text),
+                openTextDocument: (workspace, documentId) => workspace.OpenDocument(documentId),
+                cancellationToken);
+        }
 
-            // Create PreviewWorkspace around the buffer to be displayed in the diff preview
-            // so that all IDE services (colorizer, squiggles etc.) light up in this buffer.
-            var leftDocumentId = DocumentId.CreateNewId(document.Project.Id);
-            var leftSolution = document.Project.Solution
-                .RemoveAdditionalDocument(document.Id)
-                .AddAdditionalDocument(leftDocumentId, document.Name, oldBuffer.AsTextContainer().CurrentText);
-            var leftDocument = leftSolution.GetAdditionalDocument(leftDocumentId);
-            var leftWorkspace = new PreviewWorkspace(leftSolution);
-            leftWorkspace.OpenAdditionalDocument(leftDocumentId);
+        public Task<object> CreateRemovedAdditionalDocumentPreviewViewAsync(TextDocument document, double zoomLevel, CancellationToken cancellationToken)
+        {
```

Comment:
when we have text document, is there a way to see whether the document is Document or additional document or analyzerconfigdocument?

Reply:
Current master sources: You can determine this by casting it to `Document` or `AnalyzerConfigDocument` and default is additional document, as we don't have a sub-type `AdditionalDocument`.

After this PR: `TextDocument.Kind` exposes a new enum with three values that allow you to determine this. Additionally, the PR also adds a new type `AdditionalDocument: TextDocument`, so an explicit cast can also be used. Even though latter is probably sufficient by itself, the enum is still helpful as we have lot of places in our code where we want to pass around the document ID and just the kind.